File: Blind75/linked list/linked list cycle.py
# link : https://leetcode.com/problems/linked-list-cycle/submissions/916583584/

'''
2 ways to solve
1. Hashset 
storing node as it is, it will be used as objects
memory & time = O(n)

2. 2 pointers
Memory = O(1)
Time = O(n)

slow and fast pointers
slow moves 1 step
fast moves 2 step
fast pointer will always meet slow pointer at the same time
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow, fast = head, head
        
        while fast and fast.next: # because fast pointer is swifted by 2 positions
            slow = slow.next
            fast = fast.next.next
            if slow == fast: # if they meet then cycle exists
                return True
        return False # out of bounce so no cycle

File: grind75/linked list cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        // slow and fast pointers
        // fast -> slow = cycle
        ListNode *s = head, *f = head;
        while(f != nullptr && f->next != nullptr) {
        // while(s != nullptr && f!= nullptr) { not checking f->next if its a null ptr then nullptr->next is invalid
        // Because you must ensure both f itself and its immediate next (f->next) are valid before trying to go two hops forward.
            s = s->next;
            f = f->next->next;
            if( f == s ) {
                return true;
            }
        }
        return false;
    }
};